vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1224 @@
+        if (read){
+            cn.examples_index.clear();
+            for (uint32_t k = 0; k < leaf_n_examples; k++)
+                    cn.examples_index.push_back(0);
+        }
+        for (uint32_t k = 0; k < leaf_n_examples; k++)
+            writeit(cn.examples_index[k], "example_location");
+    }
+
+    void save_load_memory_tree(memory_tree& b, io_buf& model_file, bool read, bool text)
+    {
+        stringstream msg;
+        if (model_file.files.size() > 0){
+            if (read)
+                b.test_mode = true;
+
+            if (read)
+            {
+                size_t ss = 0;
+                writeit(ss, "stride_shift");
+                b.all->weights.stride_shift(ss);
+            }
+            else
+            {
+                size_t ss = b.all->weights.stride_shift();
+                writeit(ss, "stride_shift");
+            }
+            
+            writeit(b.max_nodes, "max_nodes");
+            writeit(b.learn_at_leaf, "learn_at_leaf");
+            writeit(b.oas, "oas");
+	        //writeit(b.leaf_example_multiplier, "leaf_example_multiplier")
+            writeitvar(b.nodes.size(), "nodes", n_nodes); 
+            writeit(b.max_num_labels, "max_number_of_labels");
+
+            if (read){
+                b.nodes.clear();
+                for (uint32_t i = 0; i < n_nodes; i++)
+                    b.nodes.push_back(node());
+            }
+            
+            //node  
+            for(uint32_t i = 0; i < n_nodes; i++){
+                save_load_node(b.nodes[i], model_file, read, text, msg);
+            }
+            //deal with examples:
+            writeitvar(b.examples.size(), "examples", n_examples);
+            if (read){
+                b.examples.clear();
+                for (uint32_t i = 0; i < n_examples; i++){
+                    example* new_ec = &calloc_or_throw<example>();
+                    b.examples.push_back(new_ec);
+                }
+            }
+            for (uint32_t i = 0; i < n_examples; i++)
+                save_load_example(b.examples[i], model_file, read, text, msg, b.oas);
+	    //cout<<"done loading...."<<endl;
+            
+            
+        }
+    }
+    //////////////////////////////End of Save & Load///////////////////////////////
+} //namespace
+
+base_learner* memory_tree_setup(options_i& options, vw& all)
+{
+    using namespace memory_tree_ns;
+	auto tree = scoped_calloc_or_throw<memory_tree>();
+    option_group_definition new_options("Memory Tree");
+
+    new_options.add(make_option("memory_tree", tree->max_nodes).keep().help("Make a memory tree with at most <n> nodes"))
+        .add(make_option("max_number_of_labels",tree->max_num_labels).default_value(10).help("max number of unique label"))
+        .add(make_option("leaf_example_multiplier",tree->leaf_example_multiplier).default_value(1).help("multiplier on examples per leaf (default = log nodes)"))
+        .add(make_option("Alpha", tree->alpha).default_value(0.1f).help("Alpha"))
```

Comment:
This should be lower-case to be consistent with the other reductions that take an alpha parameter.

```suggestion
        .add(make_option("alpha", tree->alpha).default_value(0.1f).help("Alpha"))
```